### PR TITLE
WIP: Submodule Refactor Phase 1

### PR DESCRIPTION
**Pull request by DarinJ**
This is going to have to be broken into chunks.

-1. Pulled out mapreduce, storm and spark (sparkstreaming in spark) responders.
-2. Added methods to ResponderPlugin interface to setup a responder for a test and return the decrypted response.
-2. Refactored `org.apache.pirk.test.DistributedTestSuite`, `org.apache.pirk.test.Utils.BaseTests` to handle submodules.  This will likely need more work as the unit tests in pirk-core depend on some of the code in `org.apache.pirk.test.Utils.BaseTests` as a goal is to pull the distributed tests into a submodule.

I've tested the Distributed Tests with the JSON input format on a local K8s hadoop cluster.  I haven't been able to set up an Elastic Search instance yet, but it looks like those tests "Fail correctly".

Todo Pull out Querier, Responder, Benchmarks, etc Modules as mentioned on the mailing list.  Some methods in the framework responder impls need pulled out into a either an abstract class, interface with default methods, or class with static methods.

This is still rough, but I think it's the right direction.  I took notes on the DistributedTestSuite and how state is being moved to various locations via system configurations.  I'll post those and suggestions to the mailing list.  Likely it'll be useful to the configuration discussion.

### CODE REVIEW

File: pirk-core/src/main/java/org/apache/pirk/test/distributed/DistributedTestCLI.java [hunk 1]
```diff
@@ -18,12 +18,7 @@
  */
 package org.apache.pirk.test.distributed;
 
-import org.apache.commons.cli.CommandLine;
-import org.apache.commons.cli.CommandLineParser;
-import org.apache.commons.cli.GnuParser;
-import org.apache.commons.cli.HelpFormatter;
-import org.apache.commons.cli.Option;
-import org.apache.commons.cli.Options;
+import org.apache.commons.cli.*;
```
Comment (ellisonanne) [inline pirk-core/src/main/java/org/apache/pirk/test/distributed/DistributedTestCLI.java]: We need to take out the * imports - prob just inadvertently added in via your IDE - this is throughout the PR

File: pirk-core/src/main/java/org/apache/pirk/test/distributed/testsuite/DistTestSuite.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    BaseTests.testDNSHostnameQuery(dataElements, fs, platform, true, 1);
+    BaseTests.testDNSIPQuery(dataElements, fs, platform, true, 1);
+
+    SystemConfiguration.setProperty("pirTest.embedSelector", "false");
+    BaseTests.testDNSHostnameQuery(dataElements, fs, platform, true, 2);
+    BaseTests.testDNSIPQuery(dataElements, fs, platform, true, 2);
+
+    BaseTests.testSRCIPQueryNoFilter(dataElements, fs, platform, true, 2);
+
+    // Test hit limits per selector
```
Comment (ellisonanne) [inline pirk-core/src/main/java/org/apache/pirk/test/distributed/testsuite/DistTestSuite.java]: Nitpick - duplicate line

File: pirk-core/src/main/java/org/apache/pirk/test/distributed/testsuite/DistTestSuite.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    BaseTests.testDNSNXDOMAINQuery(dataElements, fs, SPARK, true, 3);
+    SystemConfiguration.setProperty("pirTest.embedSelector", "false");
+    BaseTests.testDNSNXDOMAINQuery(dataElements, fs, SPARK, true, 3);
+
+    logger.info("Completed testESInputSpark");
+  }
+
+  public static void testSparkStreaming(FileSystem fs, List<JSONObject> pirDataElements) throws Exception
+  {
+    // testJSONInputSparkStreaming(fs, pirDataElements);
```
Comment (ellisonanne) [inline pirk-core/src/main/java/org/apache/pirk/test/distributed/testsuite/DistTestSuite.java]: What's going on here?

File: pirk-mapreduce/src/main/java/org/apache/pirk/responder/wideskies/mapreduce/MapReduceResponder.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    try
+    {
+      return File.createTempFile("finalResultsFile", ".txt");
+    } catch (IOException e)
+    {
+      throw new PIRException("Failed to open temp file: finalResultsFile.txt." + e.getMessage());
+    }
+  }
+
+  protected void writeDecriptedResults(DecryptResponse decryptResponse, File output, int numThreads) throws PIRException
```
Comment (ellisonanne) [inline pirk-mapreduce/src/main/java/org/apache/pirk/responder/wideskies/mapreduce/MapReduceResponder.java]: Nitpick - decrypted is misspelled... ;)

File: pirk-spark/src/main/java/org/apache/pirk/responder/wideskies/spark/SparkResponder.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    try
+    {
+      return File.createTempFile("finalResultsFile", ".txt");
+    } catch (IOException e)
+    {
+      throw new PIRException("Failed to open temp file: finalResultsFile.txt." + e.getMessage());
+    }
+  }
+
+  protected void writeDecriptedResults(DecryptResponse decryptResponse, File output, int numThreads) throws PIRException
```
Comment (ellisonanne) [inline pirk-spark/src/main/java/org/apache/pirk/responder/wideskies/spark/SparkResponder.java]: Decrypted misspelled

File: pirk-spark/src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/SparkStreamingResponder.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    try
+    {
+      return File.createTempFile("finalResultsFile", ".txt");
+    } catch (IOException e)
+    {
+      throw new PIRException("Failed to open temp file: finalResultsFile.txt." + e.getMessage());
+    }
+  }
+
+  protected void writeDecriptedResults(DecryptResponse decryptResponse, File output, int numThreads) throws PIRException
```
Comment (ellisonanne) [inline pirk-spark/src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/SparkStreamingResponder.java]: Decrypted misspelled

**Comment by ellisonanne** [comment]
Looks like it's headed in the right direction - a few minor comments inline. I did not run the tests.

Before the PR is completed, the LICENSE and NOTICE files need to be appropriately added to each submodule.

File: pirk-core/src/main/java/org/apache/pirk/test/distributed/DistributedTestCLI.java (hunk 1 above)
Comment (DarinJ) [inline pirk-core/src/main/java/org/apache/pirk/test/distributed/DistributedTestCLI.java]: yep, will do.

File: pirk-core/src/main/java/org/apache/pirk/test/distributed/testsuite/DistTestSuite.java (hunk 3 above)
Comment (DarinJ) [inline pirk-core/src/main/java/org/apache/pirk/test/distributed/testsuite/DistTestSuite.java]: That was some intermediate work, test* functions are now dead code.  I generalized testJSONInput(String platform, FileSystem fs, List<JSONObject> dataElements and testESInput(String platform, FileSystem fs, List<JSONObject> dataElements).

I didn't delete just yet as I was using for reference.  Will do in final PR.

**Comment by DarinJ** [comment]
Still need to take care of some * imports and the LICENSE/NOTICE files.

**Comment by DarinJ** [comment]
Closing to change from WIP

**Pull request closed without merging**